File: src/main/java/com/github/kindrat/cassandra/client/ui/window/editor/main/table/DataTableView.java
package com.github.kindrat.cassandra.client.ui.window.editor.main.table;

import com.github.kindrat.cassandra.client.ui.DataObject;
import com.github.kindrat.cassandra.client.ui.eventhandler.TableClickEvent;
import com.github.kindrat.cassandra.client.ui.keylistener.TableCellCopyHandler;
import javafx.geometry.HPos;
import javafx.geometry.VPos;
import javafx.scene.control.SelectionMode;
import javafx.scene.control.TableColumn;
import javafx.scene.control.TableView;
import javafx.scene.layout.GridPane;

import java.util.List;

public class DataTableView extends TableView<DataObject> {
    public DataTableView() {
        setEditable(true);
        setPrefSize(200, 200);
        GridPane.setHalignment(this, HPos.CENTER);
        GridPane.setValignment(this, VPos.CENTER);
        GridPane.setRowIndex(this, 1);
        setColumnResizePolicy(TableView.UNCONSTRAINED_RESIZE_POLICY);

        TableViewSelectionModel<DataObject> selectionModel = getSelectionModel();
        selectionModel.setSelectionMode(SelectionMode.SINGLE);
        selectionModel.setCellSelectionEnabled(true);

        setOnMouseClicked(new TableClickEvent<>(this));
    }

    public void setDataColumns(List<TableColumn<DataObject, Object>> columns) {
        getColumns().clear();
        getItems().clear();
        setEditable(true);
        getColumns().addAll(columns);
    }

    public TableCellCopyHandler buildCopyHandler() {
        return new TableCellCopyHandler(this);
    }
}

File: src/main/kotlin/com/github/kindrat/cassandra/client/ui/window/editor/main/table/DataTableView.kt
package com.github.kindrat.cassandra.client.ui.window.editor.main.table

import com.datastax.oss.driver.api.core.metadata.schema.TableMetadata
import com.github.kindrat.cassandra.client.ui.DataObject
import com.github.kindrat.cassandra.client.ui.eventhandler.TableClickEventHandler
import com.github.kindrat.cassandra.client.ui.keylistener.TableCellCopyHandler
import javafx.geometry.HPos
import javafx.geometry.VPos
import javafx.scene.control.SelectionMode
import javafx.scene.control.TableColumn
import javafx.scene.control.TableView
import javafx.scene.layout.GridPane

class DataTableView : TableView<DataObject>() {
    fun setDataColumns(columns: List<TableColumn<DataObject, Any>>) {
        getColumns().clear()
        items.clear()
        isEditable = true
        getColumns().addAll(columns)
    }

    fun onTableSelected(tableMetadata: TableMetadata) {
        onMouseClicked = TableClickEventHandler(this)
    }

    fun buildCopyHandler(): TableCellCopyHandler {
        return TableCellCopyHandler(this)
    }

    init {
        isEditable = true
        setPrefSize(200.0, 200.0)
        GridPane.setHalignment(this, HPos.CENTER)
        GridPane.setValignment(this, VPos.CENTER)
        GridPane.setRowIndex(this, 1)
        columnResizePolicy = UNCONSTRAINED_RESIZE_POLICY
        val selectionModel = selectionModel
        selectionModel.selectionMode = SelectionMode.SINGLE
        selectionModel.isCellSelectionEnabled = true
    }
}